File: MATLAB - LIQUAC copy/DResolution.m
% The purpose of this script is to give you an exact tie line, xE, and xR.
% The script can be very slow and should only be used after you have guesses at
% D from LIQUAC.m


% GLOBAL VARIABLES
% You need to instantiate certain variables as global so that they remain
% constant and can be used in the function below.
global r q MW valency z_exp_total xE_exp xR_exp T_exp_total ...
    rho_solvent_total dielec_solvent_total z_exp_ternary_total D_values ...    
    ip_UNIQUAC objLR objMR objSR objFlash salt ...
    ip_Gmehling b_ij c_ij b_jcja c_jcja ...
    z_exp_ternary T_exp rho_solvent dielec_solvent



% INPUT VALUES -- USER-SUPPLIED
% All relevant data needs to be saved in the following file format:
% "/Users/username/Desktop/salt-solvent/parameters.xlsx"
% See LIQUACinputs.m for additional details
salt = "NaCl";
solvent = "DIPA";
username = "kinnarishah";



% INPUT VALUES -- GRABBED FROM USER-SUPPLIED VIA CSV FILES
% See LIQUACinputs.m for additional details on each variable
obj = LIQUACinputs;
[r, q, MW, valency] = obj.speciesData(salt, solvent, username);
[z_exp_total, xE_exp, xR_exp, T_exp_total, rho_solvent_total, dielec_solvent_total, ...
    Selec_exp, ip_Gmehling] = obj.experimentalData(salt, solvent, username);
z_exp_ternary_total = [z_exp_total(:,1) ./ sum(z_exp_total(:,1:3),2), z_exp_total(:,2) ./ sum(z_exp_total(:,1:3),2), z_exp_total(:,3) ./ sum(z_exp_total(:,1:3),2)];
z_exp_ternary_total = round(z_exp_ternary_total,5);



% INPUT VALUES -- USER SUPPLIED
% This script runs on the basis that you have a good guess for the tie line
% slopes based on your interaction paramters. You must save the D_values 
% from LIQUAC.m and supply them here to make this script work.
D_values = readtable(strcat("/Users/",username,"/Desktop/",salt,"-",solvent,"/D_values.csv"));
D_values = D_values{:,:};



% UNIQUAC PARAMETERS FOR SOLVENT-WATER
% DIPA-H2O values are currently supplied; use UNIQUAC_Gibbsfits.m if you
% want to determine the UNIQUAC parameters for a different solvent-water
% pair. Note that it is most robust to solve for these separately and hold
% them constant during LIQUAC fitting
ip_UNIQUAC = [-2.13881699866569,	-2.13155946322135,	541.616375907126,	587.067085535208];



% OBJECT INITIALIZATION
% Initializes objects for each component of LIQUAC: Long, Medium, and Short
% Range. Also initializes GibbsMinimization.m which solves for equilibrium
% mole fractions and tie line slopes
objLR = LongRange;
objMR = MediumRange;
objSR = ShortRange; 
objFlash = GibbsMinimization;



% LIQUAC INTERACTION PARAMETERS TO BE FITTED: [b_ij, c_ij]
% Look to Gmehling group papers for good starting guesses
interaction_params = [-1.49245819676001,	1.99466864109731,	0.526147597939412,	0.705754761227743];



% Holds interaction parameters constant. The purpose of this script is to
% find the equilibrium mole fractions and tie line slopes, not to iterate
% on the interaction parameters.
b_ij = [interaction_params(1:2), ip_Gmehling(1:2)];
c_ij = [interaction_params(3:4), ip_Gmehling(3:4)];
b_jcja = ip_Gmehling(5);
c_jcja = ip_Gmehling(6);
ip = [b_ij, c_ij, b_jcja, c_jcja, ip_UNIQUAC];


% Initializes xE, xR, and tie line slopes (D_refined) all of which will be
% solved for
xE = zeros(size(z_exp_total,1), size(z_exp_total,2));
xR = zeros(size(z_exp_total,1), size(z_exp_total,2));
D_refined = zeros(size(z_exp_total,1),1);


% Solves for xE, xR, D_refined
for i = 1:size(z_exp_total,1)

    % z denotes feed mole fractions
    z_exp_ternary = z_exp_ternary_total(i,:);
    T_exp = T_exp_total(i,:);
    rho_solvent = rho_solvent_total(i,:);
    dielec_solvent = dielec_solvent_total(i,:);

    [D_refined(i), f_val2, exitflag2, output2] = fminsearch(@findD, D_values(i));
    [xE(i,:), xR(i,:)] = objFlash.Gibbs_LIQUAC_withD(z_exp_ternary, r, q, T_exp, ip, MW, valency, rho_solvent, dielec_solvent, salt, objMR, objSR, objLR, D_refined(i));
    
end



z_exp_ternary = z_exp_ternary_total;
T_exp = T_exp_total;
rho_solvent = rho_solvent_total;
dielec_solvent = dielec_solvent_total;

xE_ternary = xE(:,1:3) ./ sum(xE(:,1:3),2);
xR_ternary = xR(:,1:3) ./ sum(xR(:,1:3),2);



% Same function as is in GibbsMinimization.m; see that script for fully
% commented details
function max_area = findD(D)

    global r q MW valency T_exp rho_solvent dielec_solvent ...
        ip_UNIQUAC objLR objMR objSR salt z_exp_ternary ...
        b_ij c_ij b_jcja c_jcja

    sigfigs = 5;

    if D > 0
        x_NaCl_min = -(z_exp_ternary(2)/D - z_exp_ternary(3));
        x_NaCl_max = (D * z_exp_ternary(3) + 1 - z_exp_ternary(2)) / (1 + D);
    else
        x_NaCl_min = (D * z_exp_ternary(3) + 1 - z_exp_ternary(2)) / (1 + D);
        x_NaCl_max = z_exp_ternary(3) - z_exp_ternary(2) / D;
    end    
    
    if (x_NaCl_min <= 0)
        x_NaCl_min = 10^-sigfigs;
    end
    if (x_NaCl_max > 1)
        x_NaCl_max = 1;
    end
    x_NaCl_min = round(x_NaCl_min, sigfigs);
    x_NaCl_max = round(x_NaCl_max, sigfigs);
    
    x_NaCl = transpose(x_NaCl_min:(10^-sigfigs):(x_NaCl_max - (10^-sigfigs)));
    node_lengths = zeros(size(x_NaCl,1),1) + 10^-sigfigs;
    index = 1:size(x_NaCl,1);

    y_H2O = -D * (x_NaCl_max - x_NaCl) + (1 - x_NaCl_max);
    z_solv = 1 - x_NaCl - y_H2O;
    x_tot = [z_solv, y_H2O, x_NaCl, x_NaCl];
    x_tot = x_tot ./ sum(x_tot,2);

    temperature = zeros(size(x_NaCl,1),1) + T_exp;
    rho_solv = zeros(size(x_NaCl,1),1) + rho_solvent;
    dielec_solv = zeros(size(x_NaCl,1),1) + dielec_solvent;

    ln_gamma_LR = objLR.func_LR(MW, valency, x_tot, temperature, rho_solv, dielec_solv, salt);
    ln_gamma_MR = objMR.func_MR(b_ij, c_ij, b_jcja, c_jcja, x_tot, MW, valency, salt);
    ln_gamma_SR = objSR.func_SR(x_tot, r, q, temperature, ip_UNIQUAC, MW);
    ln_gamma = ln_gamma_LR + ln_gamma_MR + ln_gamma_SR; 
    ln_gamma(:,3:4) = [abs(valency(3)) / (abs(valency(3)) + abs(valency(4))) * ln_gamma(:,3) + abs(valency(4)) / (abs(valency(3)) + abs(valency(4))) * ln_gamma(:,4), ...
                    abs(valency(3)) / (abs(valency(3)) + abs(valency(4))) * ln_gamma(:,3) + abs(valency(4)) / (abs(valency(3)) + abs(valency(4))) * ln_gamma(:,4)];
                
    
    dg_mix = x_tot(:,1) .* log(x_tot(:,1)) + x_tot(:,2) .* log(x_tot(:,2)) + ...
        x_tot(:,3) .* log(x_tot(:,3)) + x_tot(:,1) .* ln_gamma(:,1) + ...
        x_tot(:,2) .* ln_gamma(:,2) + x_tot(:,3) .* ln_gamma(:,3) + ...
        x_tot(:,4) .* log(x_tot(:,4)) + x_tot(:,4) .* ln_gamma(:,4);


    if isreal(dg_mix)
            
        m1 = (dg_mix(size(x_NaCl,1)) - dg_mix(1)) / (x_NaCl(size(x_NaCl,1)) - x_NaCl(1));
        b1 = dg_mix(1) - m1 * x_NaCl(1);
        dg_line = m1 * x_NaCl + b1;
        deltas = dg_mix - dg_line;
        indices = index(islocalmin(deltas));

        if (size(indices, 2) ~= 2)
            indices = index(islocalmin(dg_mix));
        end

        if size(indices,2) == 2

            guesses = indices(1:2);

            trap_area = abs(dg_mix(guesses(1)) + dg_mix(guesses(2))) * abs(x_NaCl(guesses(1)) - x_NaCl(guesses(2))) / 2;
            curve_area = abs(sum(dg_mix(guesses(1):guesses(2)) .* node_lengths((guesses(1):guesses(2)))));
            max_area = sqrt(1 + D^2) * (trap_area - curve_area);
    
                                
            for j = 1:(guesses(1) - 1)
    
                trap_area = abs(dg_mix(guesses(2)) + dg_mix(j)) * abs(x_NaCl(guesses(2)) - x_NaCl(j)) / 2;
                curve_area = abs(sum(dg_mix(j:guesses(2)) .* node_lengths(j:guesses(2))));
                difference = sqrt(1 + D^2) * (trap_area - curve_area);
    
                if difference > max_area
                    max_area = difference;
                    indices(1) = j;
                elseif difference < max_area
                    break
                end
            end
    
            for j = guesses(1):(guesses(2) - 1)
    
                trap_area = abs(dg_mix(guesses(2)) + dg_mix(j)) * abs(x_NaCl(guesses(2)) - x_NaCl(j)) / 2;
                curve_area = abs(sum(dg_mix(j:guesses(2)) .* node_lengths(j:guesses(2))));
                difference = sqrt(1 + D^2) * (trap_area - curve_area);
    
                if difference > max_area
                    max_area = difference;
                    indices(1) = j;
                elseif difference < max_area
                    break
                end
            end
            

            for k = (guesses(2) + 1):size(x_NaCl,1)
                
                trap_area = abs(dg_mix(k) + dg_mix(indices(1))) * abs(x_NaCl(k) - x_NaCl(indices(1))) / 2;
                curve_area = abs(sum(dg_mix(indices(1):k) .* node_lengths(indices(1):k)));
                difference = sqrt(1 + D^2) * (trap_area - curve_area);
                
                if difference > max_area
                    max_area = difference;
                    indices(2) = k;
                elseif difference < max_area
                    break
                end
            end

            
            for k = (indices(1) + 1):(guesses(2) - 1)
                trap_area = abs(dg_mix(k) + dg_mix(indices(1))) * abs(x_NaCl(k) - x_NaCl(indices(1))) / 2;
                curve_area = abs(sum(dg_mix(indices(1):k) .* node_lengths(indices(1):k)));
                difference = sqrt(1 + D^2) * (trap_area - curve_area);
                
                if difference > max_area
                    max_area = difference;
                    indices(2) = k;
                elseif difference < max_area
                    break
                end
            end
            max_area = -max_area;
        else
            max_area = 10^10;
        end
    else
        max_area = 10^10;
    end

end

File: d_resolution.py
"""
d_resolution.py
===============
Tie-line refinement script — equivalent to MATLAB DResolution.m.

PURPOSE
-------
After running liquac_fit.py you have coarse estimates of the tie-line
slopes (D_values.csv).  This script takes those as starting guesses and
uses scipy.optimize.minimize (Nelder-Mead) to find the D value that
maximises the Eubanks area for each data point individually, giving
higher-precision xE and xR.

WORKFLOW
--------
1. Run liquac_fit.py first to produce D_values.csv.
2. Edit the USER SETTINGS block below.
3. Run:
       python d_resolution.py
4. Results are saved to:
       <data_dir>/<salt>-<solvent>/results_Dresolution.csv
"""

from pathlib import Path

import numpy as np
import pandas as pd
from scipy.optimize import minimize

from liquac_inputs       import LIQUACInputs
from LongRange          import LongRange
from MediumRange        import MediumRange
from ShortRange         import ShortRange
from gibbs_minimization  import GibbsMinimization, _build_scan, _compute_ln_gamma, _dg_mix, _eubanks_area


# ═══════════════════════════════════════════════════════════════════════════════
# USER SETTINGS
# ═══════════════════════════════════════════════════════════════════════════════

SALT    = "NaCl"
SOLVENT = "DIPA"
DATA_DIR = "/Users/lucascaldentey/Desktop/Yip Lab"

# UNIQUAC parameters (same as used during fitting — hold fixed)
IP_UNIQUAC = np.array([-2.13881699866569, -2.13155946322135,
                        541.616375907126,  587.067085535208])

# Fitted LIQUAC interaction parameters from liquac_fit.py
INTERACTION_PARAMS = np.array([-1.49245819676001,  1.99466864109731,
                                 0.52614759793941,   0.70575476122774])

# ═══════════════════════════════════════════════════════════════════════════════


def _find_D_objective(D_scalar, z_ternary, T, rho_s, diel_s,
                      b_ij, c_ij, b_jcja, c_jcja, ip_UNIQUAC,
                      r, q, MW, valency, salt,
                      objLR, objMR, objSR):
    """
    Returns the *negative* Eubanks area for a given D (so minimising this
    maximises the area).  Returns 1e10 if the scan is degenerate.

    Equivalent to the MATLAB `findD` nested function in DResolution.m.
    """
    D = float(D_scalar)
    sigfigs = 5

    scan = _build_scan(z_ternary, D, sigfigs)
    if scan is None:
        return 1e10

    x_NaCl, node_lengths, x_tot = scan

    ln_gamma = _compute_ln_gamma(
        x_tot, T, rho_s, diel_s,
        MW, valency, r, q,
        b_ij, c_ij, b_jcja, c_jcja, ip_UNIQUAC,
        salt, objLR, objMR, objSR,
    )
    dg = _dg_mix(x_tot, ln_gamma)

    if not np.all(np.isreal(dg)):
        return 1e10

    D_scale = np.sqrt(1 + D**2)
    max_area, idx_E, idx_R = _eubanks_area(dg, x_NaCl, node_lengths, D_scale)

    if idx_E < 0:
        return 1e10

    return -max_area   # negate: scipy minimises, we want to maximise area


def run_resolution():
    # ── load data ──────────────────────────────────────────────────────────────
    loader = LIQUACInputs(DATA_DIR)
    r, q, MW, valency = loader.species_data(SALT, SOLVENT)
    (z_exp, xE_exp, xR_exp, T_exp,
     rho_solvent, dielec_solvent,
     _, ip_Gmehling) = loader.experimental_data(SALT, SOLVENT)

    z_ternary = loader.to_ternary(z_exp)
    N = z_exp.shape[0]

    # ── load D starting guesses ───────────────────────────────────────────────
    d_path = Path(DATA_DIR) / f"{SALT}-{SOLVENT}" / "D_values.csv"
    if not d_path.exists():
        raise FileNotFoundError(
            f"D_values.csv not found at {d_path}.\n"
            "Run liquac_fit.py first to generate it."
        )
    D_guess = pd.read_csv(d_path, header=None).values.ravel()

    # ── assemble full interaction parameter vector ─────────────────────────────
    b_ij   = np.concatenate([INTERACTION_PARAMS[0:2], ip_Gmehling[0:2]])
    c_ij   = np.concatenate([INTERACTION_PARAMS[2:4], ip_Gmehling[2:4]])
    b_jcja = ip_Gmehling[4]
    c_jcja = ip_Gmehling[5]
    ip_full = np.concatenate([b_ij, c_ij, [b_jcja, c_jcja], IP_UNIQUAC])

    # ── initialise objects ────────────────────────────────────────────────────
    objLR = LongRange()
    objMR = MediumRange()
    objSR = ShortRange()
    objG  = GibbsMinimization()

    # ── refine each data point ────────────────────────────────────────────────
    D_refined = np.zeros(N)
    xE_pred   = np.zeros((N, 4))
    xR_pred   = np.zeros((N, 4))

    print(f"\n{'='*60}")
    print(f"D-resolution:  {SALT} / {SOLVENT}   ({N} data points)")
    print(f"{'='*60}\n")

    for i in range(N):
        z_t   = z_ternary[i]
        T_i   = T_exp[i]
        rho_i = rho_solvent[i]
        die_i = dielec_solvent[i]

        obj_fn = lambda D_arr: _find_D_objective(
            D_arr[0], z_t, T_i, rho_i, die_i,
            b_ij, c_ij, b_jcja, c_jcja, IP_UNIQUAC,
            r, q, MW, valency, SALT, objLR, objMR, objSR
        )

        res = minimize(
            obj_fn,
            x0  = [D_guess[i]],
            method = "Nelder-Mead",
            options = {"xatol": 1e-6, "fatol": 1e-8, "maxiter": 5000},
        )

        D_refined[i] = res.x[0]

        xE_pred[i], xR_pred[i] = objG.gibbs_liquac_with_D(
            z_t, r, q, T_i, ip_full, MW, valency,
            rho_i, die_i, SALT, objMR, objSR, objLR, D_refined[i]
        )

        print(f"  [{i+1:3d}/{N}]  D_guess={D_guess[i]:.4f}  "
              f"D_refined={D_refined[i]:.6f}  "
              f"f={res.fun:.6e}  converged={res.success}")

    # ── ternary normalised compositions ──────────────────────────────────────
    xE_ternary = xE_pred[:, 0:3] / xE_pred[:, 0:3].sum(axis=1, keepdims=True)
    xR_ternary = xR_pred[:, 0:3] / xR_pred[:, 0:3].sum(axis=1, keepdims=True)

    # ── save results ─────────────────────────────────────────────────────────
    out_dir = Path(DATA_DIR) / f"{SALT}-{SOLVENT}"
    cols_4   = ["solvent", "water", "cation", "anion"]
    cols_3   = ["solvent", "water", "salt"]

    df = pd.DataFrame({
        **{f"xE_{c}":         xE_pred[:, j]    for j, c in enumerate(cols_4)},
        **{f"xR_{c}":         xR_pred[:, j]    for j, c in enumerate(cols_4)},
        **{f"xE_tern_{c}":    xE_ternary[:, j] for j, c in enumerate(cols_3)},
        **{f"xR_tern_{c}":    xR_ternary[:, j] for j, c in enumerate(cols_3)},
        "D_guess":   D_guess,
        "D_refined": D_refined,
    })
    out_path = out_dir / "results_Dresolution.csv"
    df.to_csv(out_path, index=False)
    print(f"\nResults saved to: {out_path}")

    return D_refined, xE_pred, xR_pred


if __name__ == "__main__":
    run_resolution()